api/envoy/extensions/geoip_providers/maxmind/v3/maxmind.proto

```diff
@@ -0,0 +1,42 @@
+syntax = "proto3";
+
+package envoy.extensions.geoip_providers.maxmind.v3;
+
+import "envoy/extensions/geoip_providers/common/v3/common.proto";
+
+import "xds/annotations/v3/status.proto";
+
+import "udpa/annotations/status.proto";
+import "validate/validate.proto";
+
+option java_package = "io.envoyproxy.envoy.extensions.geoip_providers.maxmind.v3";
+option java_outer_classname = "MaxmindProto";
+option java_multiple_files = true;
+option go_package = "github.com/envoyproxy/go-control-plane/envoy/extensions/geoip_providers/maxmind/v3;maxmindv3";
+option (udpa.annotations.file_status).package_version_status = ACTIVE;
+option (xds.annotations.v3.file_status).work_in_progress = true;
+
+// [#protodoc-title: MaxMind Geolocation Provider]
+// MaxMind geolocation provider :ref:`configuration overview <config_geoip_providers_maxmind>`.
+// At least one geolocation database path :ref:`city_db_path <envoy_v3_api_field_extensions.geoip_providers.maxmind.v3.MaxMindConfig.city_db_path>`,
+// :ref:`isp_db_path <envoy_v3_api_field_extensions.geoip_providers.maxmind.v3.MaxMindConfig.isp_db_path>` or
+// :ref:`anon_db_path <envoy_v3_api_field_extensions.geoip_providers.maxmind.v3.MaxMindConfig.anon_db_path>` must be configured.
+// [#extension: envoy.geoip_providers.maxmind]
+
+message MaxMindConfig {
+  // Full file path to the Maxmind city database, e.g. /etc/GeoLite2-City.mmdb.
+  // Database file is expected to have .mmdb extension.
+  string city_db_path = 1 [(validate.rules).string = {pattern: "^$|^.*\\.mmdb$"}];
+
+  // Full file path to the Maxmind ASN database, e.g. /etc/GeoLite2-ASN.mmdb.
+  // Database file is expected to have .mmdb extension.
+  string isp_db_path = 2 [(validate.rules).string = {pattern: "^$|^.*\\.mmdb$"}];
+
+  // Full file path to the Maxmind anonymous IP database, e.g. /etc/GeoIP2-Anonymous-IP.mmdb.
+  // Database file is expected to have .mmdb extension.
+  string anon_db_path = 3 [(validate.rules).string = {pattern: "^$|^.*\\.mmdb$"}];
+
+  // Common provider configuration that specifies which geolocation headers will be populated with geolocation data.
+  common.v3.CommonGeoipProviderConfig common_provider_config = 4
```

Comment:
Thinking out loud here, why was this moved into the provider?
IIUC the provider defines where the data is coming from (the mmdb files above). Why does it need the geo-headers-to-add as part of the provider?

Reply:
Good question, initially i added `geo_headers_to_add` as part of geoip filter itself, but when implementing provider part had to move it to provider due to following reason:
* Provider needs to maintain a mapping of "query param (e.g. country, asn etc)->mmddb file". This should ideally be initialised once upon provider instance creation by the extension factory. If user has configured only subset of params (e.g. country, city), provider does not need to initialise 2 other mmdb databases (asn and anon) and to perform lookups in those. I did not find an easy way to reuse protobuf configuration between geoip filter and provider extension (2 different extension categories). E.g. providers are created separately from filter by a dedicated config factory. So to avoid initialising not needed databases and to skip lookups we either need to duplicate config across 2 extensions or to have this mapping in provider. Moving this logic to provider also greatly simplified geoip filter and geoip interface code.

Reply:
Apologies, I might be missing something, so bear with me while I offer a design that may not be fully correct.

IIUC there could be multiple listeners/routes with the GeoIP filter. Each one may have a different set of headers to add, and could maybe reuse the same (or different) providers.
If that's the case, I think it may be possible to decouple the "headers-to-add" and the actual providers. There could be a Provider-Manager, that handles all underlying providers, and only initializes those that are needed.
It will probably need to have 2 main methods: one to add provider configs, and one to add route-configs.
The provider config will essentially be validated and stored for ("lazy") deferred initialization, only if a relevant route-config will be present.

As a first part, I guess the manager could only support a single filter-config (not shared across other configs). It could have a single method that receives both configs (the provider and the high-level config one that contains "headers-to-add").

Let me know if I missed some optimization that could not be achieved this way.

Reply:
Thanks @adisuissa, overall design suggestion sounds good to me. It solves the problem of need for sharing part of configuration between filter and provider and offers clean api to users. I would like to keep per route configuration out of scope for now if possible, as filter is still WIP and the current PR is large in size and has been ongoing for quite some time.

Reply:
I forgot that it is still WIP, thanks for reminding me.
In that case, and as long as it is known that it will change, then LGTM on the API.

Reply:
No worries, there are remaining obligatory chunks of work, like for example support of db file reload, make encoding type configurable (maxmind only does UTF8 which may not work with H2), so filter is not fully production ready yet. Geoip headers will be moved back to filter api as part of obligatory work.